### ISSUE TITLE
add rrule in complex cases

### ISSUE THREAD

**Issue by jieli-matrix** [jieli-matrix/SparseADRules.jl#22]
All the rrules in real cases have been added and now I want to think about the complex and adjoint cases.
However, It seems `FiniteDifferences.jl` doesn't support the jv'p in complex cases. When I just initialize the nzval as random complex values, it would went wrong as follows,
``` julia
Got exception outside of a @test
  MethodError: no method matching _j′vp(::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, ::ComposedFunction{ComposedFunction{ComposedFunction{typeof(first), typeof(FiniteDifferences.to_vec)}, FiniteDifferences.var"#79#80"{ChainRulesTestUtils.var"#fnew#42"{ChainRulesTestUtils.var"#call#50"{NamedTuple{(), Tuple{}}}, Tuple{typeof(*), SparseMatrixCSC{ComplexF64, Int64}, Matrix{ComplexF64}}, Tuple{Bool, Bool, Bool}}}}, FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{var"#SparseMatrixCSC_from_vec#1"{SparseMatrixCSC{ComplexF64, Int64}, SparseMatrixCSC{ComplexF64, Int64}}, FiniteDifferences.var"#Array_from_vec#34"{Matrix{ComplexF64}, FiniteDifferences.var"#Vector_from_vec#32"{Vector{ComplexF64}, Vector{FiniteDifferences.var"#Complex_from_vec#21"}, Vector{Vector{Float64}}}}}}}, ::Vector{Float64}, ::Vector{ComplexF64})
  Closest candidates are:
    _j′vp(::Any, ::Any, ::Vector{var"#s16"} where var"#s16"<:Real, ::Vector{var"#s15"} where var"#s15"<:Real) at /Users/lijie/.julia/packages/FiniteDifferences/W3rQO/src/grad.jl:78
  Stacktrace:
```

So I just try another way: concat the real part and imag part as nzval, related codes are as follows:

``` julia
# ChainRulesTestUtils requires some method override to support sparse array
# BEGIN ChainRulesTestUtils

# we use this to generate Tangent and pass to test_rrule
# for example: A ⊢ sprand_tangent(A)
function sprand_tangent(A::SparseMatrixCSC{Tv}) where Tv<:Real
    Ā = copy(A)
    Ā.nzval .= rand(eltype(A), length(A.nzval))
    return Tangent{SparseMatrixCSC}(Ā)
end

function sprand_tangent(A::SparseMatrixCSC{Tv}) where Tv<:Complex
    Ā = SparseMatrixCSC(
        A.m,
        A.n,
        A.colptr,
        A.rowval,
        rand(real(Tv), 2*length(A.nzval))
    )
    return Tangent{SparseMatrixCSC}(Ā)
end

function Base.:(+)(a::P, b::Tangent{P}) where P<:SparseMatrixCSC
    b[1] .+ a
    return b
end

function FiniteDifferences.to_vec(A::SparseMatrixCSC{Tv}) where Tv<:Real
    v = A.nzval
    function SparseMatrixCSC_from_vec(v)
        m = A.m
        n = A.n
        colptr = A.colptr
        rowval = A.rowval
        SparseMatrixCSC(m, n, colptr, rowval, v)
    end
    return v, SparseMatrixCSC_from_vec
end

function FiniteDifferences.to_vec(A::SparseMatrixCSC{Tv}) where Tv<:Complex
    v = collect(reinterpret(real(Tv), A.nzval))
    function SparseMatrixCSC_from_vec(v)
        m = A.m
        n = A.n
        colptr = A.colptr
        rowval = A.rowval
        SparseMatrixCSC(m, n, colptr, rowval, v)
    end
    return v, SparseMatrixCSC_from_vec
end
# END ChainRulesTestUtils
```

and then test it 

``` julia
@testset "test rrule" begin
    @testset "matrix multiplication" begin
        # sparse * vector
        for T in (Float64, ComplexF64)
            A = sprand(T, 5, 3, 0.5)
            B = rand(T, 3)
            test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
            # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.

            # sparse * dense matrix
            A = sprand(T, 5, 3, 0.5)
            B = rand(T, 3, 2)
            test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
            # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.
        end 
    end
end
```
Complex cases went wrong with error messages as follows:

``` julia
test_rrule: * on SparseMatrixCSC{ComplexF64, Int64},Vector{ComplexF64}: Test Failed at /Users/lijie/.julia/packages/ChainRulesTestUtils/ugS4W/src/check_result.jl:24
  Expression: isapprox(actual, expected; kwargs...)
   Evaluated: isapprox(
           ⋅                     ⋅                   ⋅    
 0.0389629-0.364324im            ⋅           1.36762-2.9693im
 -0.696108-0.37008im             ⋅                   ⋅    
  0.370505-0.627416im  -0.385663-0.815565im  4.92051-4.24894im
  0.542053-0.102961im   0.271014-0.627069im  4.87641+0.674131im, 
   ⋅           ⋅         ⋅ 
  0.0389629    ⋅        1.36762
 -0.696108     ⋅         ⋅ 
  0.370505   -0.385663  4.92051
  0.542053    0.271014  4.87641; rtol = 1.0e-9, atol = 1.0e-9)
Stacktrace:
 [1] test_approx(actual::SparseMatrixCSC{ComplexF64, Int64}, expected::SparseMatrixCSC{Float64, Int64}, msg::String; kwargs::Base.Iterators.Pairs{Symbol, Float64, Tuple{Symbol, Symbol}, NamedTuple{(:rtol, :atol), Tuple{Float64, Float64}}})
   @ ChainRulesTestUtils ~/.julia/packages/ChainRulesTestUtils/ugS4W/src/check_result.jl:24
 [2] macro expansion
   @ ~/.julia/packages/ChainRulesTestUtils/ugS4W/src/testers.jl:238 [inlined]
 [3] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [4] test_rrule(::ChainRulesTestUtils.ADviaRuleConfig, ::typeof(*), ::ChainRulesTestUtils.PrimalAndTangent{SparseMatrixCSC{ComplexF64, Int64}, Tangent{SparseMatrixCSC, Tuple{SparseMatrixCSC{Float64, Int64}}}}, ::Vararg{Any, N} where N; output_tangent::ChainRulesTestUtils.Auto, check_thunked_output_tangent::Bool, fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, rrule_f::Function, check_inferred::Bool, fkwargs::NamedTuple{(), Tuple{}}, rtol::Float64, atol::Float64, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ChainRulesTestUtils ~/.julia/packages/ChainRulesTestUtils/ugS4W/src/testers.jl:194
```

The real part is totally right. Here I want to metion that the initilization may be wrong:
``` julia
function sprand_tangent(A::SparseMatrixCSC{Tv}) where Tv<:Complex
    Ā = SparseMatrixCSC(
        A.m,
        A.n,
        A.colptr,
        A.rowval,
        rand(real(Tv), 2*length(A.nzval))
    )
    return Tangent{SparseMatrixCSC}(Ā)
end
```
it just return `Ā.nzval `  same with the `real(Ā.nzval)`. That is the reason why the real part is totally right. We should find a way to insert imag part into gradient and convert the expected type into complex SparseCSC (must be same with the primal type).

cc: @GiggleLiu @johnnychen94

**Comment by johnnychen94** [comment]
> It seems FiniteDifferences.jl doesn't support the jv'p in complex cases.

I would suggest investigating this instead. Have you tried `jv'p` on complex inputs and verify your statement? If `jv'p` errors for Complex inputs we should try to fix it upstream. If it's just one or two missing glue methods we can add them here just like I did in #18.

**Comment by jieli-matrix** [comment]
> > It seems FiniteDifferences.jl doesn't support the jv'p in complex cases.
> 
> I would suggest investigating this instead. Have you tried `jv'p` on complex inputs and verify your statement? If `jv'p` errors for Complex inputs we should try to fix it upstream. 

Yes, I have tried it. And I have read the source code of FiniteDifferences, it only support `AbstractFloat` in numerical derivative. And only `Real` type is supported in both `jvp` and `jv'p`.

For `jvp`, its implementation and declaration [here](https://github.com/JuliaDiff/FiniteDifferences.jl/blob/8a7cd3b1d614c32fb9cddc3ddcb54c353a03ea75/src/grad.jl#L45).
For `jv'p`, its implementation and declaration [here](https://github.com/JuliaDiff/FiniteDifferences.jl/blob/8a7cd3b1d614c32fb9cddc3ddcb54c353a03ea75/src/grad.jl#L45)

**Comment by johnnychen94** [comment]
Looks like we need to provide a test-only `jvp` implementation for Complex inputs

https://github.com/JuliaDiff/FiniteDifferences.jl/blob/8a7cd3b1d614c32fb9cddc3ddcb54c353a03ea75/src/grad.jl#L55-L59

For test purposes, it's probably okay to use `jacobian` to implement this. Looks like `jacobian` supports Complex inputs (see [their test cases](https://github.com/JuliaDiff/FiniteDifferences.jl/blob/8a7cd3b1d614c32fb9cddc3ddcb54c353a03ea75/test/grad.jl#L159-L168))

**Comment by johnnychen94** [comment]
Another recommended path is to bring Complex support for `jvp` and `jv'p` in upstream FiniteDifferencePackage, but that might need more efforts than this PR.

**Comment by jieli-matrix** [comment]
> Another recommended path is to bring Complex support for `jvp` and `jv'p` in upstream FiniteDifferencePackage, but that might need more efforts than this PR.

Already solved by #24 , So I closed this issue. Thanks for johnny's useful advices and discussions!
